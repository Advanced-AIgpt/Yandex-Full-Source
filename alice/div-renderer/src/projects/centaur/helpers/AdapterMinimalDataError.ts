export class AdapterMinimalDataError extends Error {
    constructor(missedData: string[], template: string) {
        super(`Missed data (${missedData.join(', ')}) in template ${template}`);
        this.name = 'AdapterMinimalDataError';
    }
}
