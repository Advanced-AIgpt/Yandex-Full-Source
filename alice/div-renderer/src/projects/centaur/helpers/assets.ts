export function getS3Asset(path: string) {
    return `https://dialogs.s3.yandex.net/smart_displays/${path}`;
}

export function getStaticS3Asset(path: string) {
    return `https://yastatic.net/s3/dialogs/smart_displays/${path}`;
}
