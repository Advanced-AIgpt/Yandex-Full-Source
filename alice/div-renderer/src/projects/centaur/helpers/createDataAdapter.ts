import Ajv, { AnySchema, JSONSchemaType } from 'ajv';
import { DeepNonNullable, DeepRequired } from 'ts-essentials';
import { IRequestState } from '../../../common/types/common';
import { ExpFlags } from '../expFlags';

export const ajv = new Ajv();

type DataAdapter<D, R> = (data: D, requestState: IRequestState) => R;

export const createDataAdapter = <D, R>(schema: AnySchema, dataAdapter: DataAdapter<D, R>) => {
    const validate = ajv.compile(schema);

    const resultDataAdapter: DataAdapter<D, R> = (data, requestState) => {
        const isDataValid = validate(data);

        if (isDataValid === false) {
            // TODO: CENTAUR-1243: log errors

            if (requestState.hasExperiment(ExpFlags.errorOnAssert)) {
                const [firstError] = validate.errors ?? [];

                throw new Error(`DataValidationError: "${firstError.instancePath}" ${firstError.message}`);
            }
        }

        return dataAdapter(data, requestState);
    };

    return resultDataAdapter;
};

export type MakeClear<T> = DeepRequired<DeepNonNullable<T>>;
export type MakeTestSchemaType<T> = JSONSchemaType<MakeClear<T>>;
