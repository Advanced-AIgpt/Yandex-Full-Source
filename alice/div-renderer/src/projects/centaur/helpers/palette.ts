export interface ColorSet {
    light: string;
    dark: string;
}

export interface ColorScheme {
    [k: string]: ColorSet | ColorScheme;
}

export type ColorNamespace<T extends ColorScheme> = {
    [K in keyof T]: T[K] extends ColorScheme ? ColorNamespace<T[K]> : string;
};

/** https://a.yandex-team.ru/arcadia/portal/morda-schema/api/search/2/paletted-color.json */
export interface PaletteColor {
    name: string;
    color: string;
}

export interface Palette {
    light: PaletteColor[];
    dark: PaletteColor[];
}

function isColorSet(x: ColorSet | ColorScheme): x is ColorSet {
    return Object.keys(x).length === 2 && 'light' in x && 'dark' in x;
}

function colorSchemeToPalette({
    colorScheme,
    scope,
}: {
    colorScheme: ColorScheme;
    scope: string;
}): Palette {
    const palette: Palette = {
        dark: [],
        light: [],
    };

    const stack: {
        path: string[];
        node: ColorScheme | ColorSet;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        path: [key],
        node: value,
    }));

    while (stack.length) {
        const { path, node } = stack.pop() as {
            path: string[];
            node: ColorScheme | ColorSet;
        };

        if (isColorSet(node)) {
            const paletteName = `${scope}.${path.join('.')}`;
            palette.light.push({ name: paletteName, color: node.light });
            palette.dark.push({ name: paletteName, color: node.dark });
        } else {
            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    path: [...path, key],
                    node: value,
                })),
            );
        }
    }

    return palette;
}

function colorSchemeToNamespace<T extends ColorScheme>({
    colorScheme,
    scope,
}: {
    colorScheme: T;
    scope: string;
}): ColorNamespace<T & typeof commonColorScheme> {
    const rootNamespace: ColorNamespace<ColorScheme> = {};

    const stack: {
        path: string[];
        key: string;
        node: ColorScheme | ColorSet;
        parentNamespace: ColorNamespace<any>;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        key,
        path: [key],
        node: value,
        parentNamespace: rootNamespace,
    }));

    while (stack.length) {
        const { path, key, node, parentNamespace } = stack.pop() as {
            path: string[];
            key: string;
            node: ColorScheme | ColorSet;
            parentNamespace: ColorNamespace<any>;
        };

        if (isColorSet(node)) {
            parentNamespace[key] = `@{${scope}.${path.join('.')}}`;
        } else {
            const newNamespace: ColorNamespace<any> = {};
            parentNamespace[key] = newNamespace;

            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    key,
                    path: [...path, key],
                    node: value,
                    parentNamespace: newNamespace,
                })),
            );
        }
    }

    return rootNamespace as ColorNamespace<T & typeof commonColorScheme>;
}

export function getPalette<T extends ColorScheme, U extends boolean | undefined = undefined>({
    colorScheme,
    scope,
    extendsCommon,
}: {
    colorScheme: T;
    scope: string;
    extendsCommon?: U;
}): {
    palette: Palette;
    ns: ColorNamespace<U extends true ? T & typeof commonColorScheme : T>;
} {
    if (extendsCommon) {
        colorScheme = {
            ...commonColorScheme,
            ...colorScheme,
        };
    }

    return {
        palette: colorSchemeToPalette({ colorScheme, scope }),
        ns: colorSchemeToNamespace({ colorScheme, scope }) as ColorNamespace<U extends true ? T & typeof commonColorScheme : T>,
    };
}

const commonColorScheme = {
    text: {
        primary: {
            light: '#000000',
            dark: '#ffffff',
        },
    },
};
