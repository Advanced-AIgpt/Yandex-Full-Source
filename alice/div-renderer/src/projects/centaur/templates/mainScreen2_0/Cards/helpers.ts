export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
