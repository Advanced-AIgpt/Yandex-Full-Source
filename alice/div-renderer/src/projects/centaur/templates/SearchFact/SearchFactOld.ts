import {
    ContainerBlock,
    FixedSize,
    ImageBlock,
    MatchParentSize,
    TemplateCard,
    Templates,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { compact, upperFirst } from 'lodash';
import { NAlice } from '../../../../protos';
import { SuggestsBlock } from '../suggests';
import { getAsrTextFromInput, MMRequest } from '../../../../common/helpers/MMRequest';
import {
    colorWhiteOpacity50,
    colorWhiteOpacity90,
    offsetFromEdgeOfScreen,
    offsetTopWithSearchItem, simpleBackground,
} from '../../style/constants';
import { sourceHost } from '../../components/sourceHost';
import {
    text40m,
    text40r,
    title32m,
    title44r,
    title48m,
    title56m,
    title60m,
    title64m,
    title88m,
} from '../../style/Text/Text';
import EmptyDiv from '../../components/EmptyDiv';
import { CloseButtonWrapper } from '../../components/CloseButtonWrapper/CloseButtonWrapper';
import { textBreakpoint } from '../../helpers/helpers';

export const wrapFactsText = textBreakpoint(
    [90, title64m],
    [130, title60m],
    [175, title56m],
    [240, title48m],
    [280, title44r],
    [Number.POSITIVE_INFINITY, text40r],
);

const header = (t: string) => {
    return new TextBlock({
        ...text40m,
        text_color: colorWhiteOpacity50,
        text: upperFirst(t),
        height: new WrapContentSize(),
        width: new WrapContentSize(),
    });
};

const answer = (t: string) => {
    return new TextBlock({
        margins: {
            top: 24,
        },
        font_weight: 'medium',
        text_color: colorWhiteOpacity90,
        height: new WrapContentSize(),
        width: new WrapContentSize(),
        ...wrapFactsText(upperFirst(t)),
    });
};

const kkalAnswer = (t: string) => {
    return new TextBlock({
        ...title88m,
        text_color: colorWhiteOpacity90,
        height: new WrapContentSize(),
        width: new WrapContentSize(),
        text: t,
    });
};

const image = (url: string) => {
    return new ImageBlock({
        image_url: url,
        width: new FixedSize({ value: 120 }),
        height: new FixedSize({ value: 120 }),
        border: { corner_radius: 90 },
    });
};

const title = (t: string) => {
    return new TextBlock({
        ...title32m,
        text: t,
        margins: {
            right: 16,
        },
        height: new WrapContentSize(),
        width: new WrapContentSize(),
    });
};

const simpleFact = (
    {
        Question,
        Text,
        Title,
        Hostname,
    }: NAlice.NData.ITSearchFactData,
    mmRequest: MMRequest,
) => {
    const TextRequest = getAsrTextFromInput(mmRequest) ?? Question;
    return {
        items: compact([
            new EmptyDiv({
                width: new MatchParentSize(),
                height: new MatchParentSize({
                    weight: 1,
                }),
            }),
            new ContainerBlock({
                paddings: {
                    left: offsetFromEdgeOfScreen,
                    right: offsetFromEdgeOfScreen,
                },
                margins: {
                    bottom: 20,
                },
                items: compact([
                    TextRequest && header(TextRequest),
                    Text && answer(Text),
                    (Title || Hostname) &&
                    new ContainerBlock({
                        orientation: 'horizontal',
                        content_alignment_vertical: 'bottom',
                        margins: {
                            top: 16,
                        },
                        items: compact([
                            Title && title(Title),
                            Hostname && sourceHost(Hostname),
                        ]),
                    }),
                ]),
            }),
            new EmptyDiv({
                width: new MatchParentSize(),
                height: new MatchParentSize({
                    weight: 1,
                }),
            }),
            SuggestsBlock(mmRequest.ScenarioResponseBody, {
                paddings: {
                    left: offsetFromEdgeOfScreen,
                    right: offsetFromEdgeOfScreen,
                },
            }),
        ]),
    };
};

type factComponent = (
    searchFactData: NAlice.NData.ITSearchFactData,
    mmRequest: MMRequest
) => Partial<ConstructorParameters<typeof ContainerBlock>[0]>;

const extendedFactNutritionalValue: factComponent = (
    {
        Text,
        Hostname,
        Image,
        SerpData,
    }: NAlice.NData.ITSearchFactData,
    mmRequest: MMRequest,
) => {
    const infoString = SerpData?.Question && `${SerpData?.Question} - Пищевая ценность`;
    return {
        items: compact([
            new ContainerBlock({
                orientation: 'horizontal',
                items: compact([
                    Image && image(Image),
                    new TextBlock({
                        text: ' ',
                        width: new FixedSize({ value: 24 }),
                    }),
                    new ContainerBlock({
                        orientation: 'horizontal',
                        items: compact([
                            Text && kkalAnswer(Text),
                            SuggestsBlock(mmRequest.ScenarioResponseBody),
                        ]),
                    }),
                ]),
            }),
            infoString && new TextBlock({
                ...text40r,
                text: infoString,
                margins: {
                    top: 22,
                },
            }),
            Hostname && sourceHost(Hostname, {
                margins: {
                    top: infoString ? 16 : 22,
                },
            }),
        ]),
        paddings: {
            left: offsetFromEdgeOfScreen,
            right: offsetFromEdgeOfScreen,
            top: offsetTopWithSearchItem,
            bottom: offsetFromEdgeOfScreen,
        },
        content_alignment_vertical: 'center',
    };
};

const factsTypes: { [name: string]: factComponent } = {
    calories_fact: extendedFactNutritionalValue,
};

const fact = (
    searchFactData: NAlice.NData.ITSearchFactData,
    mmRequest: MMRequest,
) => {
    const factComponent = searchFactData.SnippetType && typeof factsTypes[searchFactData.SnippetType] !== 'undefined' ?
        factsTypes[searchFactData.SnippetType] : simpleFact;
    return factComponent(
        searchFactData,
        mmRequest,
    );
};

export default function SearchFactOld(
    searchFactData: NAlice.NData.ITSearchFactData,
    mmRequest: MMRequest,
) {
    return new TemplateCard(new Templates({}), {
        log_id: 'fact_card',
        states: [
            {
                state_id: 0,
                div: CloseButtonWrapper({
                    div: new ContainerBlock({
                        background: simpleBackground,
                        height: new MatchParentSize(),
                        width: new MatchParentSize(),
                        paddings: {
                            top: offsetTopWithSearchItem,
                            bottom: offsetFromEdgeOfScreen,
                        },
                        ...fact(searchFactData, mmRequest),
                    } as ConstructorParameters<typeof ContainerBlock>[0]),
                }),
            },
        ],
    });
}
