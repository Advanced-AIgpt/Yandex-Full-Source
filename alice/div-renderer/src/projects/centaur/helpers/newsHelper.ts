import { getStaticS3Asset } from './assets';

const TOPICS = new Set([
    'auto',
    'business',
    'computers',
    'culture',
    'incident',
    'index',
    'koronavirus',
    'personal',
    'politics',
    'science',
    'society',
    'sport',
    'world',
]);

const ALIASES = {
    nplus1: 'science',
} as { [name: string]: string };

export function getImageByTopic(topic: string | null | undefined): string {
    let resultImage = 'news/index.png';

    if (topic) {
        if (Object.prototype.hasOwnProperty.call(ALIASES, topic)) {
            topic = ALIASES[topic];
        }

        if (TOPICS.has(topic)) {
            resultImage = `news/${topic}.png`;
        }
    }
    return getStaticS3Asset(resultImage);
}
