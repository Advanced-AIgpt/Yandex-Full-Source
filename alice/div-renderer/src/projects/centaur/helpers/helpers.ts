/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
