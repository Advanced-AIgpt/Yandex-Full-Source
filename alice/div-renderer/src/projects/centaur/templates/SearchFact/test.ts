import { NAlice } from '../../../../protos';
import SearchFactOld from './SearchFactOld';
import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';
import { MMRequest } from '../../../../common/helpers/MMRequest';

jest.mock('../../../../common/logger');

describe('Facts scenario', () => {
    it('matches snapshot', () => {
        const mmRequest = new MMRequest({}, {}, {});

        expect(
            AnonymizeDataForSnapshot(SearchFactOld(
                {
                    Text: '69 лет',
                    SearchUrl: 'https://yandex.ru/search/touch/?l10n=ru-RU&lr=213&query_source=alice&text=%D1%81%D0%BA%D0%BE%D0%BB%D1%8C%D0%BA%D0%BE%20%D0%BB%D0%B5%D1%82%20%D0%BF%D1%83%D1%82%D0%B8%D0%BD%D1%83',
                    Title: 'Возраст',
                    SerpData: {},
                    Hostname: 'https://wikipedia.org',
                },
                mmRequest,
            )),
        ).toMatchSnapshot();
    });

    it('matches snapshot with utterance', () => {
        const mmRequestInputWithTextReq = new NAlice.NScenarios.TInput;
        mmRequestInputWithTextReq.Voice = new NAlice.NScenarios.TInput.TVoice();
        mmRequestInputWithTextReq.Voice.Utterance = 'Сколько лет Путину?';
        const mmRequest = new MMRequest({}, mmRequestInputWithTextReq, {});

        expect(AnonymizeDataForSnapshot(SearchFactOld({
            Text: '',
            SearchUrl: 'https://yandex.ru/search/touch/?l10n=ru-RU&lr=213&query_source=alice&text=%D1%81%D0%BA%D0%BE%D0%BB%D1%8C%D0%BA%D0%BE%20%D0%BB%D0%B5%D1%82%20%D0%BF%D1%83%D1%82%D0%B8%D0%BD%D1%83',
            Title: 'Возраст',
            SerpData: {},
            Hostname: 'https://wikipedia.org',
        }, mmRequest))).toMatchSnapshot();
    });
});
