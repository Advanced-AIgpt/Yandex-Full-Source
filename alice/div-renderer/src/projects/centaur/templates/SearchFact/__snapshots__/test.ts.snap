// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facts scenario matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "fact_card",
    "states": Array [
      Object {
        "div": Object {
          "height": Object {
            "type": "match_parent",
            "weight": 1,
          },
          "items": Array [
            Object {
              "background": Array [
                Object {
                  "color": "#121316",
                  "type": "solid",
                },
              ],
              "height": Object {
                "type": "match_parent",
              },
              "items": Array [
                Object {
                  "delimiter_style": Object {
                    "color": "#0fff",
                  },
                  "height": Object {
                    "type": "match_parent",
                    "weight": 1,
                  },
                  "type": "separator",
                  "width": Object {
                    "type": "match_parent",
                  },
                },
                Object {
                  "items": Array [
                    Object {
                      "font_size": 64,
                      "font_weight": "medium",
                      "height": Object {
                        "type": "wrap_content",
                      },
                      "line_height": 76,
                      "margins": Object {
                        "top": 24,
                      },
                      "text": "69 лет",
                      "text_color": "#fff",
                      "type": "text",
                      "width": Object {
                        "type": "wrap_content",
                      },
                    },
                    Object {
                      "content_alignment_vertical": "bottom",
                      "items": Array [
                        Object {
                          "font_size": 32,
                          "font_weight": "medium",
                          "height": Object {
                            "type": "wrap_content",
                          },
                          "line_height": 40,
                          "margins": Object {
                            "right": 16,
                          },
                          "text": "Возраст",
                          "text_color": "#fff",
                          "type": "text",
                          "width": Object {
                            "type": "wrap_content",
                          },
                        },
                        Object {
                          "font_size": 32,
                          "font_weight": "medium",
                          "height": Object {
                            "type": "wrap_content",
                          },
                          "line_height": 40,
                          "text": "wikipedia.org",
                          "text_color": "#80ffffff",
                          "type": "text",
                          "width": Object {
                            "type": "wrap_content",
                          },
                        },
                      ],
                      "margins": Object {
                        "top": 16,
                      },
                      "orientation": "horizontal",
                      "type": "container",
                    },
                  ],
                  "margins": Object {
                    "bottom": 20,
                  },
                  "paddings": Object {
                    "left": 48,
                    "right": 48,
                  },
                  "type": "container",
                },
                Object {
                  "delimiter_style": Object {
                    "color": "#0fff",
                  },
                  "height": Object {
                    "type": "match_parent",
                    "weight": 1,
                  },
                  "type": "separator",
                  "width": Object {
                    "type": "match_parent",
                  },
                },
                Object {
                  "item_spacing": 24,
                  "items": Array [],
                  "paddings": Object {
                    "left": 48,
                    "right": 48,
                  },
                  "type": "gallery",
                },
              ],
              "paddings": Object {
                "bottom": 48,
                "top": 142,
              },
              "type": "container",
              "width": Object {
                "type": "match_parent",
              },
            },
            Object {
              "alignment_horizontal": "right",
              "alignment_vertical": "top",
              "items": Array [
                Object {
                  "actions": Array [
                    Object {
                      "log_id": "close_fullscreen",
                      "url": "centaur://local_command?local_commands=[%7B%22command%22%3A%22close_layer%22%2C%22layer%22%3A%22DIALOG%22%2C%22stop_conversation%22%3Atrue%7D]",
                    },
                  ],
                  "alignment_horizontal": "right",
                  "background": Array [
                    Object {
                      "color": "#1affffff",
                      "type": "solid",
                    },
                  ],
                  "border": Object {
                    "corner_radius": 20,
                  },
                  "height": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                  "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                  "margins": Object {
                    "right": 36,
                    "top": 36,
                  },
                  "paddings": Object {
                    "bottom": 22,
                    "left": 22,
                    "right": 22,
                    "top": 22,
                  },
                  "preload_required": 1,
                  "scale": "fill",
                  "type": "image",
                  "width": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                },
              ],
              "type": "container",
            },
          ],
          "orientation": "overlap",
          "type": "container",
          "width": Object {
            "type": "match_parent",
            "weight": 1,
          },
        },
        "state_id": 0,
      },
    ],
  },
  "templates": Object {},
}
`;

exports[`Facts scenario matches snapshot with utterance 1`] = `
Object {
  "card": Object {
    "log_id": "fact_card",
    "states": Array [
      Object {
        "div": Object {
          "height": Object {
            "type": "match_parent",
            "weight": 1,
          },
          "items": Array [
            Object {
              "background": Array [
                Object {
                  "color": "#121316",
                  "type": "solid",
                },
              ],
              "height": Object {
                "type": "match_parent",
              },
              "items": Array [
                Object {
                  "delimiter_style": Object {
                    "color": "#0fff",
                  },
                  "height": Object {
                    "type": "match_parent",
                    "weight": 1,
                  },
                  "type": "separator",
                  "width": Object {
                    "type": "match_parent",
                  },
                },
                Object {
                  "items": Array [
                    Object {
                      "font_size": 40,
                      "font_weight": "medium",
                      "height": Object {
                        "type": "wrap_content",
                      },
                      "line_height": 56,
                      "text": "Сколько лет Путину?",
                      "text_color": "#80ffffff",
                      "type": "text",
                      "width": Object {
                        "type": "wrap_content",
                      },
                    },
                    Object {
                      "content_alignment_vertical": "bottom",
                      "items": Array [
                        Object {
                          "font_size": 32,
                          "font_weight": "medium",
                          "height": Object {
                            "type": "wrap_content",
                          },
                          "line_height": 40,
                          "margins": Object {
                            "right": 16,
                          },
                          "text": "Возраст",
                          "text_color": "#fff",
                          "type": "text",
                          "width": Object {
                            "type": "wrap_content",
                          },
                        },
                        Object {
                          "font_size": 32,
                          "font_weight": "medium",
                          "height": Object {
                            "type": "wrap_content",
                          },
                          "line_height": 40,
                          "text": "wikipedia.org",
                          "text_color": "#80ffffff",
                          "type": "text",
                          "width": Object {
                            "type": "wrap_content",
                          },
                        },
                      ],
                      "margins": Object {
                        "top": 16,
                      },
                      "orientation": "horizontal",
                      "type": "container",
                    },
                  ],
                  "margins": Object {
                    "bottom": 20,
                  },
                  "paddings": Object {
                    "left": 48,
                    "right": 48,
                  },
                  "type": "container",
                },
                Object {
                  "delimiter_style": Object {
                    "color": "#0fff",
                  },
                  "height": Object {
                    "type": "match_parent",
                    "weight": 1,
                  },
                  "type": "separator",
                  "width": Object {
                    "type": "match_parent",
                  },
                },
                Object {
                  "item_spacing": 24,
                  "items": Array [],
                  "paddings": Object {
                    "left": 48,
                    "right": 48,
                  },
                  "type": "gallery",
                },
              ],
              "paddings": Object {
                "bottom": 48,
                "top": 142,
              },
              "type": "container",
              "width": Object {
                "type": "match_parent",
              },
            },
            Object {
              "alignment_horizontal": "right",
              "alignment_vertical": "top",
              "items": Array [
                Object {
                  "actions": Array [
                    Object {
                      "log_id": "close_fullscreen",
                      "url": "centaur://local_command?local_commands=[%7B%22command%22%3A%22close_layer%22%2C%22layer%22%3A%22DIALOG%22%2C%22stop_conversation%22%3Atrue%7D]",
                    },
                  ],
                  "alignment_horizontal": "right",
                  "background": Array [
                    Object {
                      "color": "#1affffff",
                      "type": "solid",
                    },
                  ],
                  "border": Object {
                    "corner_radius": 20,
                  },
                  "height": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                  "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                  "margins": Object {
                    "right": 36,
                    "top": 36,
                  },
                  "paddings": Object {
                    "bottom": 22,
                    "left": 22,
                    "right": 22,
                    "top": 22,
                  },
                  "preload_required": 1,
                  "scale": "fill",
                  "type": "image",
                  "width": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                },
              ],
              "type": "container",
            },
          ],
          "orientation": "overlap",
          "type": "container",
          "width": Object {
            "type": "match_parent",
            "weight": 1,
          },
        },
        "state_id": 0,
      },
    ],
  },
  "templates": Object {},
}
`;
